### PR TITLE
Df mesh optimizer

### PR DESCRIPTION
**Pull request by dflanderka**

### CODE REVIEW

File: src/mesh/mesh_optimizer.hh [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+    }
+
+    inline void calculate_element_curve_values_as_zcurve_of_center() {
+        element_refs_.reserve(mesh_->n_elements());
+        for (uint i = 0; i < mesh_->n_elements(); ++i) {
+            element_refs_.emplace_back(i, zcurve_value(normalizer_.normalize(ElementAccessor<3>(mesh_, i).centre()), normalizer_.normalize(element_sizes_[i]), mesh_->element_vec_[i].dim()));
+        }
+    }
+
+    inline void calculate_element_curve_values_as_mean_of_coords() {
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:146]: Remove.

File: src/mesh/mesh_optimizer.hh [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+        element_refs_.reserve(mesh_->n_elements());
+        for (uint i = 0; i < mesh_->n_elements(); ++i) {
+            Vec3 tmp_norm = normalizer_.normalize(ElementAccessor<3>(mesh_, i).centre());
+            element_refs_.emplace_back(i, (tmp_norm[0] + tmp_norm[1] + tmp_norm[2]) / 3);
+        }
+    }
+
+    inline void calculate_element_curve_values_as_first_coord() {
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:154]: Remove.

File: src/mesh/mesh_optimizer.hh [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+        node_refs_.reserve(mesh_->n_nodes());
+        for (uint i = 0; i < mesh_->n_elements(); ++i) {
+            const Element& el = mesh_->element_vec_[element_refs_[i].original_index_];
+            for (uint j = 0; j < el.dim() + 1; ++j) {
+                node_refs_.emplace_back(el.nodes_[j], element_refs_[i].curve_value_);
+            }
+        }
+    }
+
+    inline void calculate_element_curve_values_as_mean_of_nodes() {
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:120]: Remove.

File: src/mesh/mesh_optimizer.hh [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+    inline void calculate_node_curve_values_as_mean_of_coords() {
+        node_refs_.reserve(mesh_->n_nodes());
+        for (uint i = 0; i < mesh_->n_nodes(); ++i) {
+            const Vec3 tmp_norm = normalizer_.normalize(mesh_->nodes_.vec<3>(i));
+            node_refs_.emplace_back(i, (tmp_norm[0] + tmp_norm[1] + tmp_norm[2]) / 3);
+        }
+    }
+
+    inline void calculate_node_curve_values_as_first_coord() {
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:102]: Remove.

File: src/mesh/mesh_optimizer.hh [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+    }
+
+    inline void calculate_node_curve_values_as_zcurve() {
+        node_refs_.reserve(mesh_->n_nodes());
+        for (uint i = 0; i < mesh_->n_nodes(); ++i) {
+            node_refs_.emplace_back(i, zcurve_value(normalizer_.normalize(mesh_->nodes_.vec<3>(i)), normalizer_.normalize(node_sizes_[i]), mesh_->element_vec_[i].dim()));
+        }
+    }
+
+    inline void calculate_node_curve_values_as_mean_of_coords() {
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:94]: Remove.

File: src/mesh/mesh_optimizer.hh [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+    double data[3];
+};
+
+struct Permutee {
+    inline Permutee(uint original_index, double curve_value) : original_index_(original_index), curve_value_(curve_value) {}
+    uint original_index_;
+    double curve_value_;
+};
+
+struct Normalizer {
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:38]: This should perform mapping of the mesh bounding box to the unit cube. 
Make it simple class with overloaded call operator.

File: src/mesh/mesh_optimizer.hh [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+    inline Normalizer() : shift_({0, 0, 0}), scalar_(1) {}
+    inline Normalizer(Vec3 shift, double scalar) : shift_(shift), scalar_(scalar) {}
+    inline Vec3 normalize(const Vec3 vec) {
+        return arma::vec3((vec.arma() - shift_.arma()) / scalar_);
+    }
+    inline double normalize(double size) {
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:44]: remove

File: src/mesh/mesh_optimizer.hh [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+#include <vector>
+
+#include <armadillo>
+
+#include "mesh/mesh.h"
+#include "mesh/accessors.hh"
+#include "mesh/range_wrapper.hh"
+#include "bounding_box.hh"
+
+struct Vec3 {
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:14]: Use suitable Armor::Array vector

File: src/mesh/mesh_optimizer.hh [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+                tmp_sum += node_refs_[nodeIndexes[j]].curve_value_;
+            }
+            element_refs_.emplace_back(i, tmp_sum / (mesh_->element_vec_[i].dim() + 1));
+        }
+    }
+
+    inline void calculate_element_curve_values_as_hilbert_of_centers() {
+        element_refs_.reserve(mesh_->n_elements());
+        for (uint i = 0; i < mesh_->n_elements(); ++i) {
+            element_refs_.emplace_back(i, hilbert_value(normalizer_.normalize(ElementAccessor<3>(mesh_, i).centre()), normalizer_.normalize(element_sizes_[i]), mesh_->element_vec_[i].dim()));
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:135]: Use existing element measures in hilbert and zcurve calculations, remove own calculation of element sizes.

File: src/mesh/mesh_optimizer.hh [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+        for (uint i = 0; i < node_refs_.size(); ++i) {
+            mesh_->nodes_.set(i) = nodes_backup.vec<3>(node_refs_[i].original_index_);
+            new_node_permutation[ node_refs_[i].original_index_ ] = i;
+            node_ids.add_item( mesh_->find_node_id(node_refs_[i].original_index_) );
+        }
+        mesh_->node_permutation_ = new_node_permutation;
+        mesh_->node_ids_ = node_ids;
+    }
+
+    inline void sort_elements() {
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:186]: Function for common parts of nodes and elements sorting.

File: src/mesh/mesh_optimizer.hh [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+                        return (6 + hilbert_value(2 * y - 1, 2 - 2 * z, 1 - 2 * x, 8 * eps)) / 8;
+                    } else {
+                        return (7 + hilbert_value(2 - 2 * z, 2 * x, 1 - 2 * y, 8 * eps)) / 8;
+                    }
+                }
+            }
+        }
+    }
+
+    inline double zcurve_value(double x, double eps) const {
```
Comment (jbrezmorf) [inline src/mesh/mesh_optimizer.hh:325]: Use templates, pass the point at once.

**Review by jbrezmorf — COMMENTED** [review]
- permutations must be fixed at first
- lot of space for code simplification and cleanup

File: src/io/output_msh.cc [hunk 12]
```diff
@@ -190,7 +209,18 @@ void OutputMSH::write_corner_data(OutputDataPtr output_data)
...
     file << this->offsets_->n_values() << endl; // number of values
 
-    this->write_msh_ascii_data(this->elem_ids_, output_data, true);
+    //this->write_msh_ascii_data(this->elem_ids_, output_data, true);
+    auto &id_vec = *( this->elem_ids_->get_component_data(0).get() );
+	auto &offsets_vec = *( this->offsets_->get_component_data(0).get() );
+	unsigned int n_nodes, i_corner=0;
+    for(unsigned int i=0; i < id_vec.size(); ++i) {
+        unsigned int i_gmsh_elm = output_mesh_->orig_mesh_->element_permutation(i);
+        n_nodes = (i_gmsh_elm==0) ? (offsets_vec[0]) : (offsets_vec[i_gmsh_elm]-offsets_vec[i_gmsh_elm-1]);
```
Comment (jbrezmorf) [inline src/io/output_msh.cc:218]: `offset_vec` modify to have offset_vec[0]==0 the we can simplify this code.

File: src/mesh/mesh_optimizer.hh (hunk 11 above)
Comment (jbrezmorf) [reply src/mesh/mesh_optimizer.hh:325]: Rather just keep the 3d variants.

**Comment by dflanderka** [comment]
VTK output is changed. Elements and nodes are sorted to original order (corresponds with input mesh). We don't need this modification in final version. Reference files (in ref_out dirs) will be modified and these commits will be revert:
2b55a2d, 401268f and 61cad75

File: src/mesh/mesh.h [hunk 13]
```diff
@@ -354,6 +354,18 @@ class Mesh {
         return element_ids_[pos];
     }
 
+    /// Return permutation of element on given position
+    inline unsigned int element_permutation(unsigned int pos) const
```
[the remark this answers is not in the document]
Comment (dflanderka) [reply src/mesh/mesh.h:364]: Inverzni mapa neni treba, permutace uz je delana tak, ze jde primo pouzit. Neni mi jasne, jak mapy do GMSH output predavat. V konstruktoru, specialni metodou? Mapa je nyni vytvorena jako vektor. Bylo by vhodne jej nekopirovat, ale predavat napr. referenci.

[the remark this answers is not in the document]
Comment (jbrezmorf) [reply src/mesh/mesh.h:364]: Jestli GMSHOutput vidi na OutputMesh, tak vidi i na puvodni mesh a tam muze byt primo getter pro celou tu mapu.

[the remark this answers is not in the document]
Comment (dflanderka) [reply src/mesh/mesh.h]: Pridany prislusne gettery. Opet pouzita metoda OutputMSH::write_msh_ascii_data spolecna pro vystup NodeData a ElementData - viz. vyse.

File: src/mesh/mesh_optimizer.hh (hunk 9 above)
Comment (dflanderka) [reply src/mesh/mesh_optimizer.hh:107]: Is it method ElementAccessor::measure ?

**Pull request merged**